issue/4413-wcpay-screens

nbradbury opened the pull request:
This PR adds the following WCPay onboarding screens:

1. WCPay not installed state UI - WCPayNotInstalledState
2. WCPay not activated state UI - WCPayNotActivatedState
3. Unsupported WCPay version state UI - WCPayUnsupportedVersionState
4. WCPay not setup state UI - WCPayNotSetupState

![install](https://user-images.githubusercontent.com/3903757/127160894-83c97641-9d8f-4db9-81ab-ca3be8b1ecc2.png)

~~**Note:** Currently missing from this is the dark mode illustration, but I don't think we need to hold this up for that.~~

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

nbradbury committed:
Added bare-bones WCPay fragment

nbradbury committed:
Switched to a constraint layout

nbradbury committed:
First pass at WCPay not active

nbradbury committed:
Fleshed our remaining WCPay states

nbradbury committed:
Updated layout

nbradbury committed:
Removed top margin from layout

adamzelinski reviewed (changes requested):
<img width="1379" alt="Screen Shot 2021-07-28 at 4 35 51 pm" src="https://user-images.githubusercontent.com/74938219/127275630-6183dc5a-cb90-47ed-9d55-48df8ab997e8.png">

This is looking great so far thanks @nbradbury could we please just check the title case in the top navbar matches the designs along with the header (I noticed the divider is missing)

Also, it looks like the body copy colour is a little different. Designs are showing the colour as #000000

Thanks

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/onboarding/CardReaderOnboardingFragment.kt:
```diff
@@ -83,6 +89,81 @@ class CardReaderOnboardingFragment : BaseFragment(R.layout.fragment_card_reader_
+        }
+    }
+
+    private fun showWCPayNotActivatedState(
```
I'm not sure about the best approach but my gut feeling tells me that the OnboardingViewModel I created is bit of a hybrid 🐱 🐶 . Let me try to share what's on my mind and we can perhaps discuss it and find the best solution.

We used a single layout for the whole connect card reader flow and also a single layout for the card reader payment flow. This approach bit us, since as the screens were evolving the number of differences between the steps was increasing. Re-using a single layout wasn't great there, since we had to introduce a lot of conditions - such us different margins depending on the state etc.

We decided to take another approach for the onboarding flow and create a separate state for each step of the flow. When I was writing the `OnboardingViewState` I thought we'd have a separate layout for each step.

This PR re-uses a single layout for the screens related to WCPay "errors". This make sense since all the four screens are basically identical (except of the content). However, I'm wondering  if the `OnboardingViewState` structure still makes sense since we have four states but a single layout. This forces us to copy-paste almost identical `showWCPayNotInstalledState, showWCPayUnsupportedVersionState, ...`. 

So I'd consider doing one of the following:
1. If we are concerned that the designs for the WCPay related screen will evolve and they will not be consistent in the future (that's what happened in the connection and payment flows) -> we might want to consider introducing a layout file per screen.
2. If we believe that the designs for the WCPay related screen will remain consistent even as they evolve -> we might want to consider creating a parent view state state for them, perhaps something like
```
sealed class WCPayError(
            val headerLabel: UiString,
            val hintLabel: UiString,
            val learnMoreLabel: UiString,
            @DrawableRes val illustration: Int,
            val refreshButtonLabel: UiString
        ) : OnboardingViewState(R.layout.fragment_card_reader_onboarding_wcpay) {
            abstract val refreshButtonAction: () -> Unit
            abstract val onLearnMoreActionClicked: (() -> Unit)

            data class WCPayNotInstalledState(
                override val refreshButtonAction: () -> Unit,
                override val onLearnMoreActionClicked: (() -> Unit)
            ) : WCPayError(
                headerLabel = UiString.UiStringRes(R.string.card_reader_onboarding_wcpay_not_installed_header),
                hintLabel = UiString.UiStringRes(R.string.card_reader_onboarding_wcpay_not_installed_hint),
                learnMoreLabel = UiString.UiStringRes(R.string.card_reader_onboarding_learn_more, containsHtml = true),
                illustration = R.drawable.img_woo_payments,
                refreshButtonLabel =
                UiString.UiStringRes(R.string.card_reader_onboarding_wcpay_not_installed_refresh_button)
            )
...
```
this would allow us to share the same layout and also use a single method in the fragment to map the state to the layout.


Does that make sense? What do you think about all this @nbradbury :)?

nbradbury replied:
> If we believe that the designs for the WCPay related screen will remain consistent even as they evolve -> we might want to consider creating a parent view state state for them, perhaps something like

@malinajirka I completely agree with this and had the same thought while working on this PR, but wanted to stick with the convention already set by your earlier PR. Do you want me to make this change in this PR, or separately?

malinajirka replied:
> Do you want me to make this change in this PR, or separately?

Both is fine with me ;). Thanks 🙇 !

malinajirka commented on WooCommerce/src/main/res/layout/fragment_card_reader_onboarding_wcpay.xml:
```diff
@@ -0,0 +1,60 @@
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:background="@color/color_surface"
+    android:paddingHorizontal="@dimen/major_200">
```
This limits the width of "learn more" selector. We might want to consider setting padding on each element of the layout instead of setting it on the parent. I think that's what we do on the other screens. Wdyt?

https://user-images.githubusercontent.com/2261188/127301823-cd224ab9-3556-424c-a004-d123c12b823d.mp4

malinajirka reviewed (changes requested):
Thanks @nbradbury! It looks good overall, I found just a couple visual inconsistencies and left one in-code discussion starter :).


I noticed the vertical spacing between the elements is not consistent with the  designs and also the "Unsupported country" screen. I'm not sure if it's intentional - I don't mind which approach we use, but I think it'd be great if we could make the screens consistent. Wdyt?

<img width="1293" alt="Screenshot 2021-07-28 at 11 59 51" src="https://user-images.githubusercontent.com/2261188/127303922-db3b2e09-c098-4c74-8b41-a5158bcc7bf0.png">

nbradbury committed:
Merge branch 'develop' of https://github.com/woocommerce/woocommerce-android into issue/4413-wcpay-screens

nbradbury committed:
Added dark mode payments image

nbradbury committed:
Updated WCPay layout to match country code state screen

nbradbury committed:
Fixed case on onboarding title

nbradbury commented:
@adamzelinski @malinajirka I've updated the WCPay state screens to match design and duplicate the country code state screen layout. I also updated the title case and added the toolbar divider.

![pay](https://user-images.githubusercontent.com/3903757/127373518-8f182129-caaa-49a2-acf1-f1befd6d11ec.png)

malinajirka reviewed (approved):
All looks good to me, thanks @nbradbury!

If you decide to refactor the view state as part of this PR, just let me know and I'll review the changes. Thanks ;)

nbradbury commented:
> If you decide to refactor the view state as part of this PR, just let me know and I'll review the changes. Thanks ;)

~~@malinajirka I think I'd prefer to work on that separately so this PR can be merged.~~

Scratch that, I made the changes here since this PR is still awaiting design approval.

nbradbury committed:
First pass at WCPay state refactoring

nbradbury committed:
Use the same illustration for all WCPayError states

nbradbury committed:
Removed unused param

nbradbury committed:
Removed unnecessary safe calls

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/onboarding/CardReaderOnboardingViewModel.kt:
```diff
@@ -142,69 +148,9 @@ class CardReaderOnboardingViewModel @Inject constructor(
         // TODO cardreader Update layout resource
         object WCPayAccountUnderReviewState :
-            OnboardingViewState(R.layout.fragment_card_reader_onboarding_loading) {
+            OnboardingViewState(R.layout.fragment_card_reader_onboarding_wcpay) {
```
🔍 It seems this change was not intentional, but either case it doesn't matter since `WCPayAccountUnderReviewState` is not connected to the UI yet.

malinajirka reviewed (approved):
All the changes look good to me and the app still works as expected. Great job, thanks for the improvements!

nbradbury committed:
Attempted to resolve detekt spacing issues

nbradbury committed:
Reverted accidental change to WCPayAccountUnderReviewState

nbradbury committed:
Corrected references to WCPay-related errors in tests

nbradbury committed:
Updated fragment title

nbradbury committed:
Added a divider to the onboarding fragment
